README.md
adbscan
======

A simple utility for scanning IP addresses for unprotected ADB Android devices.
Written in the Nim programming language, portable (works on most OSes and architectures out there)

Binary releases are available in GitHub Releases tab, for building from source install Nim (with nimble), clone the repository, then go to the repository directory, and run

```nimble build```


## Current status
Currently the utility can send a simple ADB connection message to the device, receive
an answer, and parse it

## Output file format
Example file output:
```
ip: 1.2.3.4 name: Hi3798MV100, model: Hi3798MV100, device: Hi3798MV100
ip: 1.2.3.4 name: marlin, model: Pixel XL, device: marlin
ip: 1.2.3.4 name: rk322x_box, model: rk322x-box, device: rk322x_box
ip: 1.2.3.4 name: aosp_noa_8g, model: TV6586_DVB, device: noa_8g
ip: 1.2.3.4 name: NV501WAC, model: NV501WAC, device: NV501WAC
ip: 1.2.3.4 name: p281, model: Hybrid 2, device: p281
ip: 1.2.3.4 name: rk322x, model: UHD-G101_V2, device: rk322x
```

For devices which aren't able to be parsed by adbscan the entry will look like
```
ip: 1.2.3.4 device info: device::http://ro.product.name =starltexx;ro.product.model=SM-G960F;ro.product.device=starlte;features=cmd,stat_v2,shell_v2
```

This program also has a simple progress bar in the terminal. 
It's not 100% representative of real "found" count since with rescans
two IPs can be scanned asynchronously.  
```
165 / 313, found 1
```

## Command line options
```
$ adbscan --help

Usage:
  cmdline [optional-params] 
Options:
  -h, --help                                 print this cligen-erated help
  --help-syntax                              advanced: prepend,plurals,..
  -i=, --input=        string     "ips.txt"  Input file
  -o=, --output=       string     "out.txt"  Output file
  -p=, --parseMode=    ParseMode  PlainText  Input file format: PlainText (default), Masscan
  -w=, --workers=      int        512        Amount of workers to use
  -r=, --rescanCount=  int        2          Amount of requests to be sent to a single IP (1 for a single scan)
```

- `PlainText` mode (which is the default) accepts input file as a line-delimited list of IPs. For example:

```
1.2.3.4
3.4.5.6
5.6.6.7
```

- `Masscan` mode accepts output of a Masscan's `-oL` output file. For example:
```
#masscan
open tcp 5555 1.2.3.4 1611375649
open tcp 5555 4.5.6.7 1611375649
# end
```

CHANGELOG.md
# Changelog

## 1.2.2
- Fix critical bug with adbscan not working on 32-bit platforms.

## 1.2.1
- Add CSV and JSON output formats
- Always include raw ADB data response in the results

## 1.2.0
- Changed the default input file mode from Masscan to PlainText.
- Changed the way rescan count is calculated.
- Changed some of the code structure in preparation for adding output formats.
- Optimized some of the code handling good results.
- Improve percentage formatting in the scanning status. 

## 1.1.0
- Made the scanning is asynchronous instead of using threads.


## 1.0.0
- Initial release.

src/adbscan.nim
import std / [
  asyncdispatch, asyncnet,
  strutils, strscans, strformat,
  options,
  tables, sequtils,
  terminal,
  random
]

import pkg / cligen

randomize()

type
  ## Message object as described per ADB protocol
  Message = object
    command: uint32
    arg0, arg1: uint32
    dataLen: uint32
    dataCrc32: uint32
    magic: uint32
  
  ## Parse mode for input files
  ParseMode = enum
    Masscan, PlainText

  OutputType = enum
    Auto, Text, Json, Csv
  
  ScanResult = object
    name: string
    model: string
    device: string
    rawData: string

const
  CmdConnect = 0x4e584e43'u32
  CmdConnectMagic = 0xb1a7b1bc'u32
  AdbVersion = 0x01000000
  MaxPayload = 4096

proc newConnectMsg(): tuple[m: Message, data: string] = 
  ## Creates a new ADB connect message
  result.data = "host::\x00"
  result.m = Message(
    command: CmdConnect,
    arg0: AdbVersion,
    arg1: MaxPayload,
    magic: CmdConnectMagic,
    dataLen: 0,
    dataCrc32: 0
  )

proc recvMessage(s: AsyncSocket): Future[Option[tuple[m: Message, data: string]]] {.async.} = 
  ## Receives an ADB Message over a socket
  var res: tuple[m: Message, data: string]
  if (await s.recvInto(addr res.m, 24)) != 24:
    # Expected 24 bytes
    return
  
  if res.m.magic != CmdConnectMagic:
    # Invalid data
    return
  
  if res.m.dataLen > 0:
    res.data = await s.recv(int(res.m.dataLen))
  
  # Verify the data checksum
  var crc = 0'u32
  for c in res.data:
    crc += uint32(ord(c))
  
  if crc != res.m.dataCrc32:
    # crc32 checksum doesn't match
    return

  result = some(res)

proc parsePayload(ip, data: string): ScanResult =
  var data = data.replace("\x00", "")
  result = ScanResult(rawData: data)
  var tmp, name, model, device: string
  const scanStr = "device::$*ro.product.name$*=$+;ro.product.model=$+;ro.product.device=$+;"

  if scanf(data, scanStr, tmp, tmp, name, model, device):
    result.name = name
    result.model = model
    result.device = device

var 
  maxWorkers = 0                      # Maximum number of allowed workers
  curWorkers = 0                      # Current amount of workers
  done = 0                            # Number of finished scans
  results: TableRef[string, ScanResult]  # Results of the scan (ip:data)

proc tryGetInfo(ip: string) {.async.} = 
  ## Checks if the Android Debug Bridge is hosted on a specified IP
  inc curWorkers
  var s = newAsyncSocket()
  
  try:
    let connectFut = s.connect(ip, Port(5555))
    # 5 seconds to connect
    if not await withTimeout(connectFut, 3500):
      return
    
    var msg = newConnectMsg()
    let sendFut = s.send(addr msg.m, sizeof(msg))
    # Send the message
    if not await withTimeout(sendFut, 2000):
      return
    
    # Wait for the reply
    let reply = s.recvMessage()
    # 3.5 seconds to receive a reply
    if not await withTimeout(reply, 3500):
      return
    
    let replyMaybe = reply.read()
    # If device is "unauthorized" then we will not get "device"
    if replyMaybe.isSome() and "device" in replyMaybe.get().data:
      results[ip] = parsePayload(ip, replyMaybe.get().data)
  
  except:
    discard
  
  finally:
    inc done
    # Yes, this is ugly but that's required
    # because isClosed doesn't work reliably
    try:
      s.close()
    except:
      discard
    dec curWorkers

proc parseFile(file: File, mode: ParseMode): seq[string] =
  ## Parses the file in specified `mode` (ParseMode)
  ##
  ## Returns a sequence of IPs
  result = @[]
  while not file.endOfFile():
    let line = file.readLine()
    if line == "": continue
    case mode
    of Masscan:
      var ip: string
      if scanf(line, "open tcp 5555 $+ ", ip):
        result.add(ip)
    of PlainText:
      result.add(line)

var allLen: int

proc updateBar() = 
  stdout.eraseLine()
  let perc = (100 * done / allLen).formatFloat(ffDecimal, 1)
  stdout.write &"{done} / {allLen} ({perc}%), found {results.len}"
  stdout.flushFile()

proc mainWork(ips: sink seq[string]) {.async.} = 
  allLen = ips.len

  # Preallocate some space
  results = newTable[string, ScanResult](allLen div 10)
  while true:
    updateBar()
    # No IPs left to scan -> exit the loop
    if ips.len == 0:
      break
    
    # While there are IPs left to scan and number of workers
    # is less than the maximum number of workers, create new workers
    while ips.len > 0 and curWorkers < maxWorkers:
      let ip = ips.pop()
      # Check if that IP is already known to have ADB running
      if ip notin results:
        asyncCheck tryGetInfo(ip)
    
    # Sleep 50ms so that we don't burn CPU cycles
    await sleepAsync(50)

proc writeText(f: File, results: TableRef[string, ScanResult]) = 
  for ip, res in results:
    f.writeLine(if res.name.len == 0:
      &"ip: {ip}, device info: {res.rawData}"
    else:
      &"ip: {ip} name: {res.name}, model: {res.model}, device: {res.device}, device info: {res.rawData}")

import std / json

proc writeJson(f: File, results: TableRef[string, ScanResult]) = 
  var root = newJArray()
  for ip, res in results:
    var obj = %*{
      "ip": ip,
      "name": res.name,
      "model": res.model,
      "device": res.device,
      "data": res.rawData
    }
    root.add obj
  f.write(root.pretty())

proc writeCsv(f: File, results: TableRef[string, ScanResult]) = 
  f.writeLine("IP,Name,Model,Device,RawData")
  for ip, res in results:
    # quote rawData since it may contain commas
    f.writeLine(&"{ip},{res.name},{res.model},{res.device},\"{res.rawData}\"")

proc cmdline(
  input = "ips.txt", output = "out.txt", 
  parseMode = PlainText, format = Auto, workers: Natural = 512, rescanCount: range[0..100] = 2
) =
  maxWorkers = workers

  var inputFile: File
  try: 
    inputFile = open(input, fmRead)
  except:
    quit(&"Cannot open {input} for reading!", 1)
  
  var outFile: File
  try:
    outFile = open(output, fmWrite)
  except:
    quit(&"Cannot open {output} for writing!", 1)

  var outType = format
  if outType == Auto:
    let tmp = output.rsplit(".", maxsplit = 1)
    if tmp.len == 2:
      let ext = tmp[1]
      case ext
      of "json": outType = Json
      of "csv": outType = Csv
      else: outType = Text
  
  var ips = parseFile(inputFile, parseMode)
  inputFile.close()

  # Add the same IPs for multiple rescans
  if rescanCount < 1:
    quit("You have to scan IPs at least once, right?")
  elif rescanCount > 1:
    ips = ips.cycle(rescanCount)
  
  # Randomize IPs
  randomize()
  ips.shuffle()

  waitFor mainWork(ips)
  
  # Wait until all requests complete (or time-out)
  while hasPendingOperations():
    poll()
  
  # Last update of the bar just for completeness
  updateBar()
  echo ""

  case outType
  of Csv: writeCsv(outfile, results)
  of Json: writeJson(outfile, results)
  of Text: writeText(outfile, results)
  else: discard

  outFile.flushFile()
  outFile.close()

proc main =
  # Use cligen to generate all command-line arguments with custom help
  dispatch(
    cmdline, 
    help = {
      "input": "Input file",
      "output": "Output file",
      "parseMode": "Input file format: PlainText (default), Masscan",
      "workers": "Amount of workers to use",
      "rescanCount": "Amount of requests to be sent to a single IP (1 for a single scan)",
      "format": "Format of the output file: Auto (based on the filename), Plain, Json, Csv"
    }
  )

main()